File: backend/agents/workItemAgent/agent.ts
import { Agent, BedrockModel } from '@strands-agents/sdk';
import { BedrockAgentCoreApp } from 'bedrock-agentcore/runtime';
import { z } from 'zod';
import { finalize_response } from './tools/agent-tools.js';
import {
  get_work_item,
  add_comment,
  add_tag,
  get_child_work_items,
  create_child_work_items,
} from './tools/azure-devops-tools.js';
import { evaluate_work_item, generate_work_items } from './tools/bedrock-tools.js';
import {
  create_incomplete_work_item_metric,
  create_work_item_generated_metric,
  create_work_item_updated_metric,
} from './tools/cloudwatch-tools.js';
import { SYSTEM_PROMPT } from './prompt.js';

// Validate required environment variables
const AWS_REGION = process.env.AWS_REGION;
if (!AWS_REGION) {
  throw new Error('AWS_REGION environment variable is required');
}
const AWS_BEDROCK_MODEL_ID = process.env.AWS_BEDROCK_MODEL_ID;
if (!AWS_BEDROCK_MODEL_ID) {
  throw new Error('AWS_BEDROCK_MODEL_ID environment variable is required');
}
const AWS_BEDROCK_KNOWLEDGE_BASE_ID = process.env.AWS_BEDROCK_KNOWLEDGE_BASE_ID;
if (!AWS_BEDROCK_KNOWLEDGE_BASE_ID) {
  throw new Error('AWS_BEDROCK_KNOWLEDGE_BASE_ID environment variable is required');
}
const RESULTS_TABLE_NAME = process.env.RESULTS_TABLE_NAME;
if (!RESULTS_TABLE_NAME) {
  throw new Error('RESULTS_TABLE_NAME environment variable is required');
}
const CONFIG_TABLE_NAME = process.env.CONFIG_TABLE_NAME;
if (!CONFIG_TABLE_NAME) {
  throw new Error('CONFIG_TABLE_NAME environment variable is required');
}
const AZURE_DEVOPS_ORGANIZATION = process.env.AZURE_DEVOPS_ORGANIZATION;
if (!AZURE_DEVOPS_ORGANIZATION) {
  throw new Error('AZURE_DEVOPS_ORGANIZATION environment variable is required');
}
const AZURE_DEVOPS_CREDENTIALS_SECRET_NAME = process.env.AZURE_DEVOPS_CREDENTIALS_SECRET_NAME;
if (!AZURE_DEVOPS_CREDENTIALS_SECRET_NAME) {
  throw new Error('AZURE_DEVOPS_CREDENTIALS_SECRET_NAME environment variable is required');
}
const AZURE_DEVOPS_SCOPE = process.env.AZURE_DEVOPS_SCOPE;
if (!AZURE_DEVOPS_SCOPE) {
  throw new Error('AZURE_DEVOPS_SCOPE environment variable is required');
}
const AZURE_DEVOPS_TENANT_ID = process.env.AZURE_DEVOPS_TENANT_ID;
if (!AZURE_DEVOPS_TENANT_ID) {
  throw new Error('AZURE_DEVOPS_TENANT_ID environment variable is required');
}
const AZURE_DEVOPS_CLIENT_ID = process.env.AZURE_DEVOPS_CLIENT_ID;
if (!AZURE_DEVOPS_CLIENT_ID) {
  throw new Error('AZURE_DEVOPS_CLIENT_ID environment variable is required');
}
const AZURE_DEVOPS_CLIENT_SECRET = process.env.AZURE_DEVOPS_CLIENT_SECRET;
if (!AZURE_DEVOPS_CLIENT_SECRET) {
  throw new Error('AZURE_DEVOPS_CLIENT_SECRET environment variable is required');
}
const FEEDBACK_FEATURE_ENABLED = process.env.FEEDBACK_FEATURE_ENABLED === 'true';
const FEEDBACK_TABLE_NAME = process.env.FEEDBACK_TABLE_NAME;
if (FEEDBACK_FEATURE_ENABLED && !FEEDBACK_TABLE_NAME) {
  throw new Error('FEEDBACK_TABLE_NAME environment variable is required when FEEDBACK_FEATURE_ENABLED is true');
}

const logger = {
  _log(level: string, message: string, extra?: Record<string, unknown>) {
    const logEntry = {
      level,
      message,
      timestamp: new Date().toISOString(),
      service: 'workItemAgent',
      ...extra,
    };
    console.log(JSON.stringify(logEntry));
  },
  info(message: string, extra?: Record<string, unknown>) {
    this._log('INFO', message, extra);
  },
  warn(message: string, extra?: Record<string, unknown>) {
    this._log('WARN', message, extra);
  },
  error(message: string, extra?: Record<string, unknown>) {
    this._log('ERROR', message, extra);
  },
};

// Create the agent with tools
const agent = new Agent({
  model: new BedrockModel({
    region: AWS_REGION,
    modelId: AWS_BEDROCK_MODEL_ID,
  }),
  systemPrompt: SYSTEM_PROMPT,
  tools: [
    evaluate_work_item,
    generate_work_items,
    finalize_response,
    get_work_item,
    add_comment,
    add_tag,
    get_child_work_items,
    create_child_work_items,
    create_incomplete_work_item_metric,
    create_work_item_generated_metric,
    create_work_item_updated_metric,
  ],
});

const app = new BedrockAgentCoreApp({
  invocationHandler: {
    requestSchema: z.object({ body: z.any() }),
    process: async (req, res: any) => {
      try {
        // Decode binary payload from AWS SDK
        const workItem = new TextDecoder().decode(req.body);
        logger.info('▶️ Decoded work item', { workItem });

        // Invoke the agent
        const response = await agent.invoke(`Here is the work item: ${workItem}`);

        logger.info('✅ Agent response', { response: response.lastMessage });

        return res.json({ response: response.lastMessage });
      } catch (err) {
        logger.error('💣 Error processing request', { error: String(err) });
        return res.status(500).json({ error: 'Internal server error' });
      }
    },
  },
});

app.run();

File: backend/agents/workItemAgent/src/agent.py
"""Work Item Agent - Python version with OpenTelemetry observability.

This agent processes Azure DevOps work items using Strands Agents SDK
with full OpenTelemetry tracing support for AWS Gen AI Observability.
"""
import os
import json
import logging
from typing import Any
from strands import Agent
from strands.models import BedrockModel
from bedrock_agentcore.runtime import BedrockAgentCoreApp
from pydantic import BaseModel
from prompt import SYSTEM_PROMPT
from tools import (
    get_work_item,
    add_comment,
    add_tag,
    get_child_work_items,
    create_child_work_items,
    evaluate_work_item,
    generate_work_items,
    create_incomplete_work_item_metric,
    create_work_item_generated_metric,
    create_work_item_updated_metric,
    finalize_response,
)

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Validate required environment variables
REQUIRED_ENV_VARS = [
    "AWS_REGION",
    "AWS_BEDROCK_MODEL_ID",
    "AWS_BEDROCK_KNOWLEDGE_BASE_ID",
    "RESULTS_TABLE_NAME",
    "CONFIG_TABLE_NAME",
    "AZURE_DEVOPS_ORGANIZATION",
    "AZURE_DEVOPS_CREDENTIALS_SECRET_NAME",
]

for var in REQUIRED_ENV_VARS:
    if not os.environ.get(var):
        raise ValueError(f"{var} environment variable is required")

AWS_REGION = os.environ["AWS_REGION"]
AWS_BEDROCK_MODEL_ID = os.environ["AWS_BEDROCK_MODEL_ID"]

# Create the BedrockAgentCoreApp
app = BedrockAgentCoreApp()

# Create the agent with tools
agent = Agent(
    model=BedrockModel(
        region_name=AWS_REGION,
        model_id=AWS_BEDROCK_MODEL_ID,
    ),
    system_prompt=SYSTEM_PROMPT,
    tools=[
        evaluate_work_item,
        generate_work_items,
        finalize_response,
        get_work_item,
        add_comment,
        add_tag,
        get_child_work_items,
        create_child_work_items,
        create_incomplete_work_item_metric,
        create_work_item_generated_metric,
        create_work_item_updated_metric,
    ],
)

class WorkItemRequest(BaseModel):
    """Request schema for the agent."""

    work_item: dict
    params: dict
    session_id: str = ""

@app.entrypoint
async def process_request(request: dict, context: Any = None) -> dict:
    """Process an incoming work item request."""
    try:
        logger.info("▶️ Received request", extra={"request": request})

        # Extract from nested body structure (matches TypeScript version)
        body = request.get("body", request)
        work_item = body.get("workItem") or body.get("work_item")
        params = body.get("params", {})
        session_id = body.get("sessionId") or body.get("session_id", "")

        logger.info(
            "Parsed work item",
            extra={"workItem": work_item, "params": params, "sessionId": session_id},
        )

        # Invoke the agent
        prompt = f"""Here is the work item and params:

Work Item: {json.dumps(work_item, indent=2)}

Params: {json.dumps(params, indent=2)}

Session ID (use as executionId): {session_id or 'not-provided'}"""

        response = await agent.invoke_async(prompt, session_id=session_id)

        # Extract text from response
        message = response.message
        
        logger.info("✅ Agent response", extra={"response": message})

        return {"response": message}

    except Exception as e:
        logger.error("🛑 Error processing request", extra={"error": str(e)})
        raise

if __name__ == "__main__":
    logger.info("Starting Work Item Agent (Python)")
    app.run()
